### PR TITLE
Get `multi.c` working in the Wasm C API

### PR DESCRIPTION
**Pull request by MarkMcCaskey**
Gets the `multi.c` example working.

This PR has significant overlap with #1949 due to changes in vec / boxed vec.

Changes in this PR:
- impl `Clone` more often, implement `*_copy` functions in terms of `Clone`
- clean up boxed vec behavior
- add `*_vec_copy` functions for boxed and unboxed versions
- fix a typo in `wasm-c-api/examples/multi.c`; fixed in the upstream Wasm C API here: https://github.com/WebAssembly/wasm-c-api/pull/163
- additional misc improvements

# Review

- [ ] Add a short description of the the change to the CHANGELOG.md file

### CODE REVIEW

File: lib/c-api/src/wasm_c_api/macros.rs [hunk 1]
```diff
@@ -191,15 +255,27 @@ macro_rules! wasm_declare_boxed_vec {
...
-                    vec.size = 0;
+            pub unsafe extern "C" fn [<wasm_ $name _vec_delete>](ptr: Option<&mut[<wasm_ $name _vec_t>]>) {
+                if let Some(vec) = ptr {
+                    if !vec.data.is_null() {
+                        let ptr: *mut Option<Box<[<wasm_ $name _t>]>> = vec.data as _;
+                        let data: Vec<Option<Box<[<wasm_ $name _t>]>>> = Vec::from_raw_parts(ptr, vec.size, vec.size);
+
+                        vec.data = ::std::ptr::null_mut();
+                        vec.size = 0;
+                    }
```
Comment (Hywan) [inline lib/c-api/src/wasm_c_api/macros.rs:278]: This conflicts with https://github.com/wasmerio/wasmer/pull/1949.

File: lib/c-api/tests/wasm-c-api/example/multi.c [hunk 2]
```diff
@@ -7,6 +7,19 @@
...
+// Use the last_error API to retrieve error messages
+void print_wasmer_error()
+{
+  int error_len = wasmer_last_error_length();
+  printf("Error len: `%d`\n", error_len);
+  char *error_str = malloc(error_len);
+  wasmer_last_error_message(error_str, error_len);
+  printf("Error str: `%s`\n", error_str);
+}
+
```
Comment (Hywan) [inline lib/c-api/tests/wasm-c-api/example/multi.c:22]: That's annoying for further updates of the `wasm-c-api` repo :-/. But I don't see how to solve this.

**Review by Hywan — COMMENTED** [review]
Looks good to me.

File: lib/c-api/tests/wasm-c-api/example/multi.c (hunk 2 above)
Comment (MarkMcCaskey) [reply lib/c-api/tests/wasm-c-api/example/multi.c:22]: Oops, that was debug code... I didn't intend to ship that -- thanks for pointing it out!

**Review by jubianchi — APPROVED** [review]

File: lib/c-api/tests/wasm-c-api/example/multi.c [hunk 3]
```diff
@@ -121,7 +121,7 @@ int main(int argc, const char* argv[]) {
   // Call.
   printf("Calling export...\n");
   wasm_val_t vals[4] = {
-    WASM_I32_VAL(1), WASM_I32_VAL(2), WASM_I32_VAL(3), WASM_I32_VAL(4)
+    WASM_I32_VAL(1), WASM_I64_VAL(2), WASM_I64_VAL(3), WASM_I32_VAL(4)
```
Comment (syrusakbary) [inline lib/c-api/tests/wasm-c-api/example/multi.c:124]: Is this intended? Did the code was updated in the wasm-c-api side?

Comment (MarkMcCaskey) [reply lib/c-api/tests/wasm-c-api/example/multi.c:124]: I linked to the PR on the Wasm C API above

**Pull request merged**